knownvalue/known_value.go

```diff
@@ -0,0 +1,10 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package knownvalue
+
+// KnownValue defines an interface that is implemented to determine equality on the basis of type and value.
+type KnownValue interface {
+	// Equal should perform equality testing.
+	Equal(other any) bool
```

Comment:
I like the idea of returning an error and the additional flexibility that this yields. 

In terms of the fully-type-named interfaces, I'm wondering about the internals. For instance, implementation of `CheckBool()` would be straightforward:

```go
func (v BoolValue) CheckBool(other bool) error {
	if other != v.value {
		return fmt.Errorf("%t does not equal %t", other, v.value)
	}

	return nil
}
```

With the call to `CheckBool()` from `expectKnownValue.CheckPlan()` looking something like the following:

```go
	switch reflect.TypeOf(result).Kind() {
	case reflect.Bool:
		v, ok := e.knownValue.(knownvalue.BoolValue)

		if !ok {
			resp.Error = fmt.Errorf("wrong type: attribute value is bool, known value type is %T", e.knownValue)

			return
		}

		err := v.CheckBool(result.(bool))

		if err != nil {
			resp.Error = fmt.Errorf("attribute value: %v does not equal expected value: %s", result, v)

			return
		}
```

However, the internals for all of the `KnownValue` types that deal with maps, or slices, would, I believe, require some internal type assertion to determine whether the fully-type-named interface methods could be called. For example:

```go
func (v ObjectValue) CheckObject(other map[string]any) error {
	if len(other) != len(v.value) {
		return fmt.Errorf("%v does not equal %v", other, v.value)
	}

	for k, v := range v.value {
		otherItem, ok := other[k]

		if !ok {
			return fmt.Errorf("%s is missing from %v", k, other)
		}

		switch reflect.TypeOf(otherItem).Kind() {
		case reflect.Bool:
			boolCheck, ok := v.(BoolValue)

			if !ok {
				return fmt.Errorf("wrong type: %T, known value type is %T", otherItem, v)
			}

			if err := boolCheck.CheckBool(otherItem.(bool)); err != nil {
				return fmt.Errorf("%v does not equal %v", otherItem, v)
			} 
		// Reflection logic for reflect.Map, reflect.Slice, reflect.String
		default:
			return fmt.Errorf("unrecognised type: %T, known value type is %T", otherItem, v)
		}
	}

	return nil
}
```

I'll go ahead and implement the `Check` interface, and perhaps we can go from there?

Reply:
In that situation, yes, I would expect the collection/object checks to need to figure out what their element or per-object-attribute check type is. Another option would be to _not_ support that functionality and leave it up to developer implementations to figure out what to do with making more generic collection/object checks that support elements/object attributes, but I think there's a lot of value in being able to offer that out of the box.

Not sure if it will make the logic easier/harder, but another way of doing this is performing a type switch on the `v.value` (presumably the per-object-attribute checks in the given example code) and raising an error if the `otherItem` doesn't conform to the given check type. This can help make sure the error messaging is tailored for the what the developer put into their object check definition rather than tailored for the data returned from Terraform, which without additional context can be harder to reason about in my opinion. e.g.

```go
// note also including the object attribute name (k from example) and not assuming
// it was an equality check error in error messaging (the checks can say so
// themselves, amongst other possible errors :) )
switch check := v.(type) {
case BoolValue:
  otherValue, ok := otherItem.(bool)

  if !ok {
    return fmt.Errorf("%s object attribute: expected bool value for BoolValue check, got: %T", k, otherItem)
  }

  if err := check.CheckBool(otherValue); err != nil {
    return fmt.Errorf("%s object attribute: %s", k, err)
  }
// ...
default:
  
}
```

Reply:
I guess a potential wrinkle here is, what if there is a dynamically typed value is being checked? Hmm. I would normally presume the value should be statically typed based on the given Terraform configuration. If not, we would be fine as long as there is some sort of escape hatch for a "I don't know the value type and I'll do all the work to figure it out" type of check, e.g. `CheckDynamic(value any) error`.

Reply:
I've refactored to use the `Check` interface:

```go
type Check interface {
  CheckValue(value any) error

  String() string
}
```

Before proceeding down the route of using fully-type-named interface, thought it might be worth considering how the `ExpectKnownValue`, `ExpectKnownOutputValue`, and `ExpectKnownOutputValueAtPath` would handle these fully-type-named interfaces.

Currently, taking `ExpectKnownValue` for instance, the type and constructor look as follows:

```go
var _ PlanCheck = expectKnownValue{}

type expectKnownValue struct {
	resourceAddress string
	attributePath   tfjsonpath.Path
	knownValue      knownvalue.Check
}

func (e expectKnownValue) CheckPlan(ctx context.Context, req CheckPlanRequest, resp *CheckPlanResponse) {
	/* ... */
}

func ExpectKnownValue(resourceAddress string, attributePath tfjsonpath.Path, knownValue knownvalue.Check) PlanCheck {
	return expectKnownValue{
		resourceAddress: resourceAddress,
		attributePath:   attributePath,
		knownValue:      knownValue,
	}
}
```

As an example, `BoolValue` currently implements the `Check` interface:

```go
var _ Check = BoolValue{}

type BoolValue struct {
	value bool
}

func (v BoolValue) CheckValue(other any) error {
	/* ... */
}

/* ... */

func BoolValueExact(value bool) BoolValue {
	return BoolValue{
		value: value,
	}
}
```

If we refactor `BoolValue` to implement the `BoolCheck` interface:

```go
type BoolCheck interface {
	CheckBool(value bool) error

	String() string
}

var _ BoolCheck = BoolValue{}

type BoolValue struct {
	value bool
}

func (v BoolValue) CheckBool(value bool) error {
	/* ... */
}

/* ... */

func BoolValueExact(value bool) BoolValue {
	return BoolValue{
		value: value,
	}
}
```

Using fully-type-named interfaces will require that we reconsider how the constructor and struct field holding the checks will be handled for `ExpectKnownValue`, `ExpectKnownOutputValue`, and `ExpectKnownOutputValueAtPath`, as we'll no longer be able to construct an `ExpectKnownValue` plan check, for instance, using the following:

```go
plancheck.ExpectKnownValue(
	"test_resource.one",
	tfjsonpath.New("bool_attribute"),
	knownvalue.BoolValueExact(true),
),
```